text/359-construct-hub-deny-list.md

```diff
@@ -0,0 +1,127 @@
+3. The "discovery" and "ingestion" lambda functions will retrieve this file for each request, and will consult the 
+   deny list for every package. If an incoming package is in the deny list, it will be skipped and a log entry will 
+   be emitted with the deny information.
+4. The "inventory" process will retrieve the deny list and will filter out all denied package from inventory. 
+   Statistics about denied package will also be included in the inventory metrics.
+5. We will add a trigger to the "catalog-builder" for object deletion so when a denied package is deleted from 
+   the packages bucket, the catalog will be rebuilt without that package.
+
+### Is this a breaking change?
+
+No.
+
+### What alternative solutions did you consider?
+
+#### Manually manage the deny list via a file on S3
+
+We considered the option of letting operators manage their deny list by directly editing a file on S3. There are multiple downsides to this approach:
+
+1. If this file gets corrupted due to a human error (e.g. invalid JSON), denied package may be ingested. 
+   To address this we would need some kind of fallback to previously good versions and this adds quite a 
+   lot of unneeded complexity. By defining the deny list in strongly-typed code, the compiler (and synthesizer) 
+   takes care of validating the input which simplifies downstream consumption of this information.
+3. Each deployment will have its own copy of the deny list and replicating the deny list across multiple 
+   environments (e.g. gamma/prod) will require additional mechanism. By allowing users to specify deny list 
+   in code, they have freedom to decide what parts of the list is replicated across environments (all/none/partial) 
+   by using simple programming techniques (e.g. constants).
+
+
+### What are the drawbacks of this solution?
+
+One benefit of the direct S3 approach is reduced SLA for blocking a package. In the proposed design, customers need to 
+commit the change to their source repository and have that change propagate through their release pipeline in order for
+the package to be removed. But since the SLA we declared for removing a package is 48 hours, this seems like a reasonable
+tradeoff. If in the future we will want to add support for "quick removal" we can always add an additional mechanism that
+will temporarily block a package.
+
+### What is the high level implementation plan?
+
+Project plan TBD.
+
+### Are there any open issues that need to be addressed later?
+
+* [ ] Packages denied in the public hub will still be allowed in private hubs until they are explicitly 
+      added to the deny list there. That might actually be the desired behavior (e.g. local deployment may 
+      want to show a package that is blocked in the public hub), but it should be communicated that if a 
+      package is reported and denied from the public hub, it will still appear in private hubs unless it is explicitly denied there.
```

Comment:
I think the documentation needs to mention that the deny-list is instance-specific. That has several implications:

- If someone files a request against one construct hub instance, only that instance may get the package deny-listed
- The "report abuse" UI should be configuration (i.e: the Construct Hub needs to show a deployment-specific email address to use for submitting reports)

Reply:
Yes, I agree with both bullets. I'll elaborate.